### ISSUE TITLE
Automatic "end" insertion when writing functions in Lua fails in some cases

### ISSUE THREAD

**Issue by RunaPhox** [mawww/kakoune#1525]
Hi. This is a problem I have whenever I try to define a table element that is  function. i.e.:
``` lua
bugdemo = {}
function bugdemo.bug()|
```
Whenever I press return in insert mode where "|" is, I end up with:
``` lua
bugdemo = {}
function bugdemo.bug()
function bugdemo.bug()end
    |
```
Instead of:
``` lua
bugdemo = {}
function bugdemo.bug()
    |
end
```
This is also true if the newline is inserted from normal mode with 'o', in the function header line, or 'O', in the line following it.
As a side note, this does not happen if the function is not in a table.
Pressing return here:
``` lua
function not_in_table()|
```
Results in:
``` lua
function not_in_table()
    |
end
```
As expected.

Edit:
I have seen this weird behavior in other places as well.
For example, if I have a comment prior to the function definition:
``` lua
-- some comment
function wat()
function wat()end
```

**Comment by alyssais** [comment]
This happens with Ruby too.

**Comment by alyssais** [comment]
Ruby reproduction script:

```
set buffer filetype ruby
exec -draft iclass Test
exec -draft o
exec -draft O
```

**Comment by alyssais** [comment]
I was able to fix the bug with this patch:

```diff
diff --git a/usr/local/share/kak/rc/base/lua.kak b/rc/base/lua.kak
index 8ef832d6..9ec31b93 100644
--- a/rc/base/lua.kak
+++ b/rc/base/lua.kak
@@ -80,7 +80,7 @@ define-command -hidden lua-indent-on-new-line %{
 define-command -hidden lua-insert-on-new-line %{
     evaluate-commands -no-hooks -draft -itersel %{
         # copy -- comment prefix and following white spaces
-        try %{ execute-keys -draft k<a-x>s^\h*\K--\h*<ret>yjp }
+        try %{ execute-keys -draft k<a-x>s^\h*\K--\h*<ret>yghjP }
         # wisely add end structure
         evaluate-commands -save-regs x %{
             try %{ execute-keys -draft k<a-x>s^\h+<ret>"xy } catch %{ reg x '' }
@@ -96,8 +96,8 @@ hook -group lua-highlight global WinSetOption filetype=lua %{ add-highlighter wi
 
 hook global WinSetOption filetype=lua %{
     hook window InsertChar .* -group lua-indent lua-indent-on-char
-    hook window InsertChar \n -group lua-indent lua-indent-on-new-line
     hook window InsertChar \n -group lua-insert lua-insert-on-new-line
+    hook window InsertChar \n -group lua-indent lua-indent-on-new-line
 
     alias window alt lua-alternative-file
 }
```

It basically reverts 5989903e7c492b90045aa8db778694d2c442429c (which broke this) and fixes #1118 in a different way that won't interfere with `end` insertion.

I propose that this change is applied to all language definitions 5989903e7c492b90045aa8db778694d2c442429c modified. @mawww would you accept such a pull request? (Before I put more time into it.)

**Comment by mawww** [comment]
If that seems to work fine I'll be happy to merge a general fix. Could you add the reproduction case you described above as a test case (probably in `test/indent/ruby`). We really lack some unit tests for indentation of all those languages.

**Comment by danr** [comment]
Relevant: #1811. It would be easier to test if the filetype plugins shared the exacty same functionality.

**Comment by alyssais** [comment]
This appears to have been fixed as a side effect of 67424aae1e00196faf303201544435b30b82b16d.

I'll PR my test case anyway to ensure it doesn't break again in future.
